### DISCUSSION TITLE
Provide skylight integration

### DISCUSSION THREAD

**Discussion by mostlyobvious** [RailsEventStore/rails_event_store#1169] [Ideas]
We already have instrumented most of the interesting internals via https://railseventstore.org/docs/instrumentation/ which for `rails_event_store` gem works with `ActiveSupport::Notifications`.

It may be possible that no further work in RES (nor in additional library) is needed. Instead we may want to document the integration process.

https://www.skylight.io/support/getting-more-from-skylight#probes

**Comment by mostlyobvious** [comment]
It seems that what we need is called Normalizers:

- https://github.com/hmcfletch/fast_jsonapi/blob/7f030a2f8b8c7136bc10a044e2577d8746d5e935/lib/fast_jsonapi/instrumentation/skylight/normalizers/serialized_json.rb
- https://github.com/skylightio/skylight-ruby/blob/v2.0.0/skylight-core/lib/skylight/core/normalizers/action_controller/process_action.rb
- https://github.com/skylightio/skylight-ruby/issues/113#issuecomment-361332713

**Comment by mostlyobvious** [comment]
https://github.com/skylightio/skylight-ruby/blob/master/skylight-core/lib/skylight/core/normalizers.rb#L51-L68

**Comment by mostlyobvious** [comment]
- normalizers have to loaded quite early, before `Skylight::Subscriber` starts, should be ok to deliver them in a gem that is required in Rails `application.rb` via Bundler
- allowed tiers:

```
TIERS = %w[
  rack
  api
  app
  view
  db
  noise
  other
].freeze
```

**Comment by mostlyobvious** [comment]
<img width="1237" alt="Zrzut ekranu 2019-07-10 o 16 51 17" src="https://user-images.githubusercontent.com/116740/60979784-99c51380-a333-11e9-9b81-af7878850645.png">

**Comment by mostlyobvious** [comment]
```ruby
class Dispatcher < ::Skylight::Core::Normalizers::Normalizer
  register "call.dispatcher.rails_event_store"

  CAT = "app.rails_event_store.dispatcher".freeze

  def normalize(_trace, _name, payload)
    [CAT, "#{payload[:subscriber]}#call", "Handle #{payload[:event].type}"]
  end
end
```

**Comment by wagenet** [comment]
@pawelpacana this looks good! If you're interested, we can integrate this into the main skylight gem.

**Comment by mostlyobvious** [comment]
@wagenet thanks! Integrating into skylight gem would be probably the best from the developer perspective.

Btw. is there any way to format the description? Like multiple lines for example?

**Comment by wagenet** [comment]
Right now we don't have a way to do it. You can file an issue at https://github.com/skylightio/skylight-ruby/issues/new and explain a bit more about what you'd like.
